File: panel/backend/internal/models/stack.go
package models

import (
	"encoding/json"
	"regexp"
	"strings"
	"time"
)

// Stack is an admin-uploaded full-stack isolated app (dashboard OR tool).
// Like one Mod is one item in Mods, one Stack is one item in Stacks: own
// frontend (spa dist | simple pages) + optional backend sidecar + own nav +
// own data. Dashboard is only category='dashboard', never the system name.
//
// Manifest/spec are opaque pass-through JSON (same contract as mods/themes):
// the backend validates only the enumerated fields below + the requested
// capabilities; the rest is stored verbatim for the frontend.
type Stack struct {
	ID          int64           `json:"id"`
	Name        string          `json:"name"`
	Slug        string          `json:"slug"`
	Category    string          `json:"category"`
	Version     string          `json:"version"`
	Description string          `json:"description"`
	Icon        string          `json:"icon"`
	Color       string          `json:"color,omitempty"`
	Runtime     string          `json:"runtime"`
	Entrypoint  string          `json:"entrypoint,omitempty"`
	Manifest    json.RawMessage `json:"manifest"`
	Spec        json.RawMessage `json:"spec,omitempty"`
	// ThemeMode is panel|custom|none: inherit the panel route theme, ship an
	// own theme.css, or run unthemed. PageStyle is spa|simple: full dist
	// bundle in an iframe, or panel-rendered markdown/html/blocks pages.
	ThemeMode string `json:"theme_mode"`
	PageStyle string `json:"page_style"`
	Active    bool   `json:"active"`
	UploadedBy *int64 `json:"uploaded_by,omitempty"`
	// OwnerID ties the stack to the uploader (migration 071 wires
	// STACKS_OWN/STACKS_ALL: Own sees owner rows only, All/umbrella sees all).
	OwnerID int64 `json:"owner_id,omitempty"`
	// OwnerName is the denormalised username joined from users.
	OwnerName string `json:"owner_name,omitempty"`
	// Source is file|url|studio|json|sample (install provenance).
	Source    string `json:"source"`
	SourceURL string `json:"source_url,omitempty"`
	// PackageSize is the .ksps byte size (0 = synthesize on download).
	PackageSize int64     `json:"package_size"`
	// ProxyPort is the loopback TCP port of the externally-run stack Go app
	// (0 = proxy off). ProxyRootURL is the first path segment the panel
	// serves that app at, e.g. 'dash' floats it at /dash/* ("" = off).
	// Migration 072.
	ProxyPort    int    `json:"proxy_port"`
	ProxyRootURL string `json:"proxy_root_url,omitempty"`
	// RemoteAddress is the node-style dial address of the stack app when it
	// runs on ANOTHER host (host:port or bare host, e.g. "10.0.0.9:7700").
	// "" means same-host: the proxy dials 127.0.0.1:ProxyPort.
	// RemoteUseTLS flips the proxy/probe between http:// and https://.
	// RemoteSkipVerify skips TLS verify for self-signed remotes.
	// Migration 076.
	RemoteAddress    string `json:"remote_address,omitempty"`
	RemoteUseTLS     bool   `json:"remote_use_tls"`
	RemoteSkipVerify bool   `json:"remote_skip_verify"`
	// TokenPrefix is the first 8 chars of the pairing token so the operator
	// can recognise which token is configured (the raw token is returned
	// only at create/rotate time, never again — mirrors nodes).
	TokenPrefix string `json:"token_prefix,omitempty"`
	// Status is "up" or "down" from heartbeat freshness. LastSeenAt is the
	// UTC timestamp of the last heartbeat (nil = never connected).
	Status     string     `json:"status"`
	LastSeenAt *time.Time `json:"last_seen_at,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// Stack install provenance codes. Matches the `stacks.source` default.
const (
	StackSourceFile   = "file"
	StackSourceURL    = "url"
	StackSourceStudio = "studio"
	StackSourceJSON   = "json"
	StackSourceSample = "sample"
)

// Stack runtimes, theme modes, page styles and categories.
const (
	StackRuntimeStatic = "static"
	StackRuntimeNodeJS = "nodejs"
	StackRuntimePython = "python"

	StackThemePanel  = "panel"
	StackThemeCustom = "custom"
	StackThemeNone   = "none"

	StackPageSPA    = "spa"
	StackPageSimple = "simple"
)

var stackSlugRe = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{0,63}$`)

// ValidStackSlug reports whether slug satisfies the panel's slug contract
// (same contract as mods: lowercase alphanumerics + hyphens, max 64).
func ValidStackSlug(slug string) bool {
	return stackSlugRe.MatchString(slug)
}

// ValidStackRuntime reports whether runtime is a known sidecar runtime.
func ValidStackRuntime(runtime string) bool {
	switch runtime {
	case StackRuntimeStatic, StackRuntimeNodeJS, StackRuntimePython:
		return true
	}
	return false
}

// ValidStackThemeMode reports whether mode is a known theme mode. Empty
// means "not declared" and normalises to panel at parse time.
func ValidStackThemeMode(mode string) bool {
	switch mode {
	case "", StackThemePanel, StackThemeCustom, StackThemeNone:
		return true
	}
	return false
}

// ValidStackPageStyle reports whether style is a known page style. Empty
// normalises to spa at parse time.
func ValidStackPageStyle(style string) bool {
	switch style {
	case "", StackPageSPA, StackPageSimple:
		return true
	}
	return false
}

var stackProxyRootRe = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{0,31}$`)

// ValidStackProxyPort reports whether port is a usable proxy target.
// 0 disables the proxy; otherwise a TCP port number is required.
func ValidStackProxyPort(port int) bool {
	return port == 0 || (port >= 1 && port <= 65535)
}

// ValidStackProxyRoot reports whether root is a usable proxy mount: one
// lowercase path segment (empty disables the proxy).
func ValidStackProxyRoot(root string) bool {
	if root == "" {
		return true
	}
	return stackProxyRootRe.MatchString(root)
}

// ReservedStackProxyRoots are first path segments the proxy mount must never
// claim: the API tree, the health probe, the favicon alias and the SPA
// asset prefix all live at the origin root beside proxy mounts.
func ReservedStackProxyRoots() []string {
	return []string{"api", "health", "favicon.ico", "assets"}
}

// IsReservedStackProxyRoot reports whether root collides with a panel-owned
// origin-root path.
func IsReservedStackProxyRoot(root string) bool {
	for _, r := range ReservedStackProxyRoots() {
		if root == r {
			return true
		}
	}
	return false
}

// StackTokenPrefix brands the stack pairing token so it is
// distinguishable from an edge token (kse_…) and an API key (ksk_…).
// The stack app operator pastes this into the app's config file.
const StackTokenPrefix = "kss_"

// ValidStackRemoteAddress reports whether addr is an acceptable node-style
// dial address for a remote stack app: host:port, bare host/hostname, or
// "" (same-host loopback — never validated here, the caller treats empty
// as "use proxy_port"). It mirrors the node address rules: no scheme,
// no whitespace, numeric ports 1..65535, bracketed IPv6 supported.
func ValidStackRemoteAddress(addr string) bool {
	addr = strings.TrimSpace(addr)
	if addr == "" {
		return true
	}
	if strings.HasPrefix(addr, "http://") || strings.HasPrefix(addr, "https://") {
		return false
	}
	if strings.ContainsAny(addr, " \t\r\n") {
		return false
	}
	if strings.HasPrefix(addr, "[") {
		end := strings.Index(addr, "]")
		if end == -1 {
			return false
		}
		rest := addr[end+1:]
		if rest == "" {
			return true
		}
		if !strings.HasPrefix(rest, ":") || rest[1:] == "" {
			return false
		}
		return validStackPortStr(rest[1:])
	}
	if strings.Count(addr, ":") > 1 {
		return true // bare IPv6 literal, no port to validate
	}
	if idx := strings.LastIndex(addr, ":"); idx >= 0 {
		if strings.TrimSpace(addr[:idx]) == "" || strings.TrimSpace(addr[idx+1:]) == "" {
			return false
		}
		return validStackPortStr(strings.TrimSpace(addr[idx+1:]))
	}
	return true
}

func validStackPortStr(p string) bool {
	if len(p) == 0 || len(p) > 5 {
		return false
	}
	n := 0
	for i := 0; i < len(p); i++ {
		if p[i] < '0' || p[i] > '9' {
			return false
		}
		n = n*10 + int(p[i]-'0')
	}
	return n >= 1 && n <= 65535
}

// IsRemoteStack reports whether the stack app lives on another host
// (RemoteAddress set) versus same-host loopback (ProxyPort only).
func (s *Stack) IsRemoteStack() bool {
	return s != nil && strings.TrimSpace(s.RemoteAddress) != ""
}

// StackDialTarget returns the scheme + authority the panel dials for
// probe/proxy: remote_address when set, else 127.0.0.1:proxyPort.
func (s *Stack) StackDialTarget() (scheme, addr string) {
	if s.IsRemoteStack() {
		scheme = "http"
		if s.RemoteUseTLS {
			scheme = "https"
		}
		return scheme, strings.TrimSpace(s.RemoteAddress)
	}
	return "http", "127.0.0.1:" + stackItoa(s.ProxyPort)
}

func stackItoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b [20]byte
	pos := len(b)
	for n > 0 {
		pos--
		b[pos] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		pos--
		b[pos] = '-'
	}
	return string(b[pos:])
}

// StackPermission is one capability a stack declared it needs. Activation is
// refused until every row has Granted == true (same gate as mods).
type StackPermission struct {
	ID          int64  `json:"id"`
	StackID     int64  `json:"stack_id"`
	Capability  string `json:"capability"`
	AccessLevel string `json:"access_level"`
	Granted     bool   `json:"granted"`
}

// Stack capability codes. Deliberately NOT reusing the mod CapXxx codes:
// stacks reach capabilities through the sidecar proxy allow-list, so the
// namespace stays separate even where the English overlaps.
const (
	StackCapMetricsRead  = "metrics.read"
	StackCapInstancesRead = "instances.read"
	StackCapKVReadWrite  = "kv.read_write"
	StackCapDBReadWrite  = "db.read_write"
	StackCapOutboundHTTP = "outbound_http"
	StackCapNotify       = "notify"
)

// AllowedStackCapabilities returns the capability codes a stack manifest may
// request. The repo rejects any other string at insert time.
func AllowedStackCapabilities() []string {
	return []string{
		StackCapMetricsRead,
		StackCapInstancesRead,
		StackCapKVReadWrite,
		StackCapDBReadWrite,
		StackCapOutboundHTTP,
		StackCapNotify,
	}
}

// StackEnv is one saved environment KEY row (secret values stay masked).
type StackEnv struct {
	Key    string `json:"key"`
	Masked bool   `json:"masked,omitempty"`
	Secret bool   `json:"secret"`
}

// StackKV is one namespaced key/value row.
type StackKV struct {
	Key       string `json:"key"`
	Value     string `json:"value"`
	UpdatedAt time.Time `json:"updated_at"`
}

File: panel/frontend/src/shared/types/stack.ts
// Frontend types for the Stacks (full-stack isolated apps) system.
// One `stack` is one item like one `mod` is one item in Mods: own frontend
// (spa dist | simple pages) + optional backend sidecar + own nav + own data.
// Keep capability codes in sync with internal/models/stack.go.

export interface StackPermission {
  id: number;
  stack_id: number;
  capability: string;
  access_level: string;
  granted: boolean;
}

export interface Stack {
  id: number;
  name: string;
  slug: string;
  category: string;
  version: string;
  description: string;
  icon: string;
  color?: string;
  runtime: string;
  entrypoint?: string;
  manifest: Record<string, any> | null;
  spec: Record<string, any> | null;
  theme_mode: 'panel' | 'custom' | 'none';
  page_style: 'spa' | 'simple';
  active: boolean;
  // App proxy (migration 072): loopback port + mount segment of the
  // externally-run stack Go app (0/"" = proxy off). Mirrors models.Stack.
  proxy_port: number;
  proxy_root_url?: string;
  owner_name?: string;
  source?: StackSource;
  source_url?: string;
  package_size?: number;
  permissions: StackPermission[];
  pending: number;
  created_at: string;
  updated_at: string;
}

export type StackSource = 'file' | 'url' | 'json' | 'sample';

export interface StackSourceMeta {
  key: StackSource;
  label: string;
  dot: string;
  badge: string;
}

export const STACK_SOURCES: StackSourceMeta[] = [
  { key: 'file', label: 'Uploaded file', dot: 'bg-sky-400', badge: 'bg-sky-900/40 text-sky-200 border-sky-700/50' },
  { key: 'url', label: 'Installed from URL', dot: 'bg-violet-400', badge: 'bg-violet-900/40 text-violet-200 border-violet-700/50' },
  { key: 'json', label: 'Posted as JSON', dot: 'bg-amber-400', badge: 'bg-amber-900/40 text-amber-200 border-amber-700/50' },
  { key: 'sample', label: 'Built-in sample', dot: 'bg-teal-400', badge: 'bg-teal-900/40 text-teal-200 border-teal-700/50' },
];

export const stackSourceMeta = (key: string | undefined): StackSourceMeta | undefined =>
  STACK_SOURCES.find((s) => s.key === key);

// Stack capability codes — mirrors models.AllowedStackCapabilities().
export const StackCapability = {
  MetricsRead: 'metrics.read',
  InstancesRead: 'instances.read',
  KVReadWrite: 'kv.read_write',
  DBReadWrite: 'db.read_write',
  OutboundHTTP: 'outbound_http',
  Notify: 'notify',
} as const;

export interface StackCapabilityMeta {
  key: string;
  label: string;
  description: string;
  dot: string;
  badge: string;
}

export const STACK_CAPABILITIES: StackCapabilityMeta[] = [
  { key: StackCapability.MetricsRead, label: 'Metrics (read)', description: 'Read panel/instance metrics for dashboards.', dot: 'bg-emerald-400', badge: 'bg-emerald-900/60 text-emerald-200 border-emerald-700/60' },
  { key: StackCapability.InstancesRead, label: 'Instances (read)', description: 'Read instance inventory/status.', dot: 'bg-sky-400', badge: 'bg-sky-900/60 text-sky-200 border-sky-700/60' },
  { key: StackCapability.KVReadWrite, label: 'Key-value storage', description: 'Read/write the stack namespaced KV store.', dot: 'bg-amber-400', badge: 'bg-amber-900/60 text-amber-200 border-amber-700/60' },
  { key: StackCapability.DBReadWrite, label: 'Database', description: 'Read/write the stack SQL tables (shared prefix).', dot: 'bg-red-400', badge: 'bg-red-900/60 text-red-200 border-red-700/60' },
  { key: StackCapability.OutboundHTTP, label: 'Outbound HTTP', description: 'Fetch external HTTP(S) APIs.', dot: 'bg-violet-400', badge: 'bg-violet-900/60 text-violet-200 border-violet-700/60' },
  { key: StackCapability.Notify, label: 'Notifications', description: 'Post panel notifications/toasts.', dot: 'bg-orange-400', badge: 'bg-orange-900/60 text-orange-200 border-orange-700/60' },
];

export const stackCapabilityMeta = (key: string): StackCapabilityMeta | undefined =>
  STACK_CAPABILITIES.find((c) => c.key === key);

export interface StackPermissionRequest {
  capability: string;
  access_level: string;
}

export interface StackActivateConflict {
  error: string;
  message: string;
  pending: number;
  permissions: StackPermission[];
}

export interface StackNavEntry {
  slug: string;
  label: string;
  icon: string;
}

export interface StackEngineStatus {
  enabled: boolean;
  stacks: { slug: string; active: boolean }[];
}

export interface StackPageEntry {
  slug: string;
  title: string;
  icon?: string;
  file: string;
}

export interface StackPageContent {
  slug: string;
  title: string;
  type: 'html' | 'markdown' | 'blocks';
  content: string;
}

export type StackThemeMode = 'panel' | 'custom' | 'none';
export type StackPageStyle = 'spa' | 'simple';

export const STACK_THEME_MODES: Array<{ value: StackThemeMode; label: string; hint: string }> = [
  { value: 'panel', label: 'Panel theme', hint: 'Inherits the panel route theme via --ks-* tokens' },
  { value: 'custom', label: 'Custom theme.css', hint: 'Ships frontend/theme.css (or frontend/dist/theme.css)' },
  { value: 'none', label: 'Unthemed', hint: 'No theme tokens, raw bundle styles only' },
];

export const STACK_PAGE_STYLES: Array<{ value: StackPageStyle; label: string; hint: string }> = [
  { value: 'spa', label: 'SPA bundle', hint: 'Full frontend/dist bundle in a sandboxed iframe' },
  { value: 'simple', label: 'Simple pages', hint: 'Panel-rendered markdown/html/blocks from frontend/pages/' },
];

export const STACK_CATEGORIES = ['dashboard', 'tool', 'tracker', 'status', 'crud', 'custom'] as const;

export const STACK_RUNTIMES = ['static', 'nodejs', 'python'] as const;

export function slugify(input: string): string {
  return input
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 64);
}

// ---- Workdir file manager ----------------------------------------------

export interface StackFileEntry {
  name: string;
  is_dir: boolean;
  size: number;
  mode?: string;
  mod_time?: number;
}

export interface StackFileContent {
  path: string;
  content: string;
  size: number;
}

// ---- Workdir file manager ----------------------------------------------

export interface StackFileEntry {
